geopyspark/geotrellis/layer.py

```diff
@@ -480,46 +504,50 @@ def collect_metadata(self, extent=None, layout=None, crs=None, tile_size=256):
             TypeError: If either ``extent`` and ``layout`` is not defined but the other is.
         """
 
-        if extent and not isinstance(extent, dict):
-            extent = extent._asdict()
-
-        if layout and not isinstance(layout, dict):
-            layout = layout._asdict()
-
         if not crs:
             crs = ""
 
         if isinstance(crs, int):
             crs = str(crs)
 
-        if extent and layout:
-            json_metadata = self.srdd.collectMetadata(extent, layout, crs)
-        elif not extent and not layout:
-            json_metadata = self.srdd.collectMetadata(str(tile_size), crs)
+        if layout:
+            json_metadata = self.srdd.collectMetadata(layout, crs)
         else:
-            raise TypeError("Could not collect metadata with {} and {}".format(extent, layout))
+            json_metadata = self.srdd.collectMetadata(str(tile_size), crs)
 
         return Metadata.from_dict(json.loads(json_metadata))
 
-    def reproject(self, target_crs, resample_method=ResampleMethod.NEAREST_NEIGHBOR):
-        """Reproject every individual raster to ``target_crs``, does not sample past tile boundary
+    def reproject(self, target_crs, layout=None, resample_method=ResampleMethod.NEAREST_NEIGHBOR):
+        """Reproject rasters to ``target_crs``.
+        When layout is ``None``, the reproject does not sample past tile boundary.
+        When layout is given the reproject will tile the rasters and produce a tiled layer.
 
         Args:
-            target_crs (str or int): The CRS to reproject to. Can either be the EPSG code,
-                well-known name, or a PROJ.4 projection string.
+            target_crs (str or int): Target CRS of reprojection.
+                Either EPSG code, well-known name, or a PROJ.4 string.
+            layout (
+                :obj:`~geopyspark.geotrellis.GlobalLayout` or
+                :obj:`~geopyspark.geotrellis.LocalLayout`, optional
+            ): Target raster layout when reprojecting.
             resample_method (str or :class:`~geopyspark.geotrellis.constants.ResampleMethod`, optional):
                 The resample method to use for the reprojection. If none is specified, then
                 ``ResampleMethods.NEAREST_NEIGHBOR`` is used.
 
         Returns:
```

Comment:
I know that we wanted to have just one method handling all cases, but do you think it might be better to split the `reproject` method into two, separate operations? One would never sample past the tiles' boundaries and the other would. That might be better than having a method return one of two things depending on its inputs.